SSA/Projects/LLVMAndRiscv.lean

```diff
@@ -0,0 +1,144 @@
+
+inductive Ty where
+  | llvm : (LLVM.Ty) -> Ty
+  | riscv : (Dialect.Ty RISCV64.RV64) -> Ty
+  deriving DecidableEq, Repr
+
+inductive Op where
+  | llvm : (Dialect.Op LLVM) -> Op
+  | riscv : (Dialect.Op RISCV64.RV64) -> Op
+  | builtin.unrealized_conversion_cast.riscvToLLVM : Op
+  | builtin.unrealized_conversion_cast.LLVMToriscv : Op
+  deriving DecidableEq, Repr
+
+/-- Semantics of an unrealized conversion cast from RISC-V 64 to LLVM.
+We wrap `BitVec 64`in `Option (BitVec 64)` -/
+def builtin.unrealized_conversion_cast.riscvToLLVM (toCast : BitVec 64 ) : Option (BitVec 64 ) := some toCast
+
+/--
+Semantics of an unrealized conversion cast from LLVM to RISC-V.
+This cast attempts to lower an `(Option (BitVec 64))` to a concrete `(BitVec 64)`.
+If the value is `some`, we extract the underlying `BitVec`.
+If it is `none` (e.g., an LLVM poison value), we default to the zero `BitVec`.
+-/
+def builtin.unrealized_conversion_cast.LLVMToriscv (toCast : Option (BitVec 64)) : BitVec 64 := toCast.getD 0#64
+
+@[simp]
+abbrev LLVMPlusRiscV : Dialect where
+  Op := Op
+  Ty := Ty
+
+instance : TyDenote (Dialect.Ty LLVMPlusRiscV) where
+  toType := fun
+    | .llvm llvmTy => TyDenote.toType llvmTy
+    | .riscv riscvTy => TyDenote.toType riscvTy
+
+@[simp]
+instance LLVMPlusRiscVSignature : DialectSignature LLVMPlusRiscV where
+  signature
+  | .llvm llvmOp => .llvm <$> DialectSignature.signature llvmOp
+  | .riscv riscvOp => .riscv <$> DialectSignature.signature riscvOp
+  | .builtin.unrealized_conversion_cast.riscvToLLVM =>
+      {sig := [Ty.riscv .bv], outTy := Ty.llvm (.bitvec 64), regSig := []}
+  | .builtin.unrealized_conversion_cast.LLVMToriscv =>
+      {sig := [Ty.llvm (.bitvec 64)], outTy := (Ty.riscv .bv), regSig := []}
+
+instance : ToString (Dialect.Ty LLVMPlusRiscV)  where
+  toString t := repr t |>.pretty
+
+@[simp_denote]
+def llvmArgsFromHybrid : {tys : List LLVM.Ty} → HVector TyDenote.toType (tys.map LLVMRiscV.Ty.llvm) → HVector TyDenote.toType tys
+  | [], .nil => .nil
+  | _ :: _, .cons x xs => .cons x (llvmArgsFromHybrid xs)
+
+@[simp_denote]
+def riscvArgsFromHybrid : {tys : List RISCV64.RV64.Ty} → HVector TyDenote.toType (tys.map LLVMRiscV.Ty.riscv) → HVector TyDenote.toType tys
+  | [], .nil => .nil
+  | _ :: _, .cons x xs => .cons x (riscvArgsFromHybrid xs)
+
+@[simp, reducible]
+instance : DialectDenote (LLVMPlusRiscV) where
+  denote
+  | .llvm (llvmOp), args , .nil => DialectDenote.denote llvmOp (llvmArgsFromHybrid args) .nil
+  | .riscv (riscvOp), args , .nil => DialectDenote.denote riscvOp (riscvArgsFromHybrid args) .nil
+  | .builtin.unrealized_conversion_cast.riscvToLLVM, elemToCast, _ =>
+    let toCast : (BitVec 64) := (elemToCast.getN 0 (by simp [DialectSignature.sig, signature])) -- adapting to the newly introduced PoisonOr wrapper.
+    let casted : Option (BitVec 64) := builtin.unrealized_conversion_cast.riscvToLLVM toCast
+    PoisonOr.ofOption casted
```

Comment:
Please have the cast semantics function return a `PoisonOr`, then you don't have to use `ofOption` here